File: Data Structures & Algorithms/trapping-rain-water/submission-0.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) {
        if (height.length <= 2) return 0;
        let maxArea = 0;
        let left = 0;
        let right = height.length - 1;
        let leftMax = height[left];
        let rightMax = height[right];

        while (left <= right) {
            if (leftMax < rightMax) {
                leftMax = Math.max(leftMax, height[left]);
                maxArea += leftMax - height[left];
                left++;
                console.log('left', leftMax, 'max', maxArea);
            } else {
                rightMax = Math.max(rightMax, height[right]);
                maxArea += rightMax - height[right];
                right--;
                console.log('right', rightMax, 'max', maxArea);
            }
        }
        return maxArea;
    }
}

File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        left, right = 0, len(heights) - 1 
        area = 0
        while left < right:
            dx = right - left
            dy = min(heights[left], heights[right])
            area = max(area, dx * dy)
            if heights[left] < heights[right]:
                left += 1
            else:
                right -= 1

        return area
            
